cpp/src/io/orc/reader_impl.cu

```diff
@@ -661,238 +880,299 @@ table_with_metadata reader::impl::read(size_type skip_rows,
+                d_dst, buffer->data(), len, cudaMemcpyHostToDevice, stream.value()));
+              stream.synchronize();
+            }
           }
-          if (_metadata->per_file_metadata[stripe_source_mapping.source_idx]
-                .source->is_device_read_preferred(len)) {
-            CUDF_EXPECTS(
-              _metadata->per_file_metadata[stripe_source_mapping.source_idx].source->device_read(
-                offset, len, d_dst, stream) == len,
-              "Unexpected discrepancy in bytes read.");
-          } else {
-            const auto buffer =
-              _metadata->per_file_metadata[stripe_source_mapping.source_idx].source->host_read(
-                offset, len);
-            CUDF_EXPECTS(buffer->size() == len, "Unexpected discrepancy in bytes read.");
-            CUDA_TRY(
-              cudaMemcpyAsync(d_dst, buffer->data(), len, cudaMemcpyHostToDevice, stream.value()));
-            stream.synchronize();
-          }
-        }
 
-        // Update chunks to reference streams pointers
-        for (size_t col_idx = 0; col_idx < num_columns; col_idx++) {
-          auto &chunk         = chunks[stripe_chunk_index * num_columns + col_idx];
-          chunk.start_row     = stripe_start_row;
-          chunk.num_rows      = stripe_info->numberOfRows;
-          chunk.encoding_kind = stripe_footer->columns[_selected_columns[col_idx]].kind;
-          chunk.type_kind     = _metadata->per_file_metadata[stripe_source_mapping.source_idx]
-                              .ff.types[_selected_columns[col_idx]]
-                              .kind;
-          chunk.decimal_scale = _metadata->per_file_metadata[stripe_source_mapping.source_idx]
-                                  .ff.types[_selected_columns[col_idx]]
-                                  .scale.value_or(0);
-          chunk.rowgroup_id = num_rowgroups;
-          chunk.dtype_len   = (column_types[col_idx].id() == type_id::STRING)
-                              ? sizeof(string_index_pair)
-                              : cudf::size_of(column_types[col_idx]);
-          if (chunk.type_kind == orc::TIMESTAMP) {
-            chunk.ts_clock_rate = to_clockrate(_timestamp_type.id());
+          // Update chunks to reference streams pointers
+          uint32_t max_num_rows = 0;
+          for (size_t col_idx = 0; col_idx < num_columns; col_idx++) {
+            auto &chunk = chunks[stripe_idx * num_columns + col_idx];
+            // start row, number of rows in a each stripe and total number of rows
+            // may change in lower levels of nesting
+            chunk.start_row = (level == 0)
+                                ? stripe_start_row
+                                : _col_meta.child_start_row[stripe_idx * num_columns + col_idx];
+            chunk.num_rows =
+              (level == 0)
+                ? stripe_info->numberOfRows
+                : _col_meta.num_child_rows_per_stripe[stripe_idx * num_columns + col_idx];
```

Comment:
same: either `num_child_rows_per_stripe` can be a vector of vectors or be accessed with a 2d span.